tools/Microsoft.AspNet.Server.Kestrel.GeneratedCode/FrameFeatureCollection.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.Dnx.Compilation.CSharp;
+using Microsoft.AspNet.Http.Features.Internal;
+using Microsoft.AspNet.Http.Features.Authentication;
+
+namespace Microsoft.AspNet.Server.Kestrel.GeneratedCode
+{
+    // This project can output the Class library as a NuGet Package.
+    // To enable this option, right-click on the project and select the Properties menu item. In the Build tab select "Produce outputs on build".
+    public class FrameFeatureCollection : ICompileModule
+    {
+        static string Each<T>(IEnumerable<T> values, Func<T, string> formatter)
+        {
+            return values.Select(formatter).Aggregate((a, b) => a + b);
+        }
+
+        public virtual void BeforeCompile(BeforeCompileContext context)
+        {
+            var syntaxTree = Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(GeneratedFile());
+            context.Compilation = context.Compilation.AddSyntaxTrees(syntaxTree);
+        }
+
+        public static string GeneratedFile()
+        {
+            var commonFeatures = new[]
+            {
+                typeof(IHttpRequestFeature),
+                typeof(IHttpResponseFeature),
+                typeof(IHttpRequestIdentifierFeature),
+                typeof(IHttpSendFileFeature),
+                typeof(IServiceProvidersFeature),
+                typeof(IHttpAuthenticationFeature),
+                typeof(IHttpRequestLifetimeFeature),
+                typeof(IQueryFeature),
+                typeof(IFormFeature),
+                typeof(IResponseCookiesFeature),
+                typeof(IItemsFeature),
+                typeof(IHttpConnectionFeature),
+                typeof(ITlsConnectionFeature),
+                typeof(IHttpUpgradeFeature),
+                typeof(IHttpWebSocketFeature),
+                typeof(ISessionFeature),
+            };
+
+            return $@"
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http 
+{{
+    public partial class Frame
+    {{
+        {Each(commonFeatures.Select((feature, index) => new { feature, index }), entry => $@"
+        private const long flag{entry.feature.Name} = {1 << entry.index};")}
+
+        {Each(commonFeatures, feature => $@"
+        private static readonly Type {feature.Name}Type = typeof(global::{feature.FullName});")}
+
+        private long _featureOverridenFlags = 0L;
```

Comment:
I don't think we need the flag, TBH, because the distinction in all cases is "null or not" on the field. Know what I mean? Because the IFeatureCollection doesn't have Count, Remove, or ContainsKey, TryGet, there's no distinction between not being set and being null.

Plus you'd still need to null out all the fields on reset anyway, because you don't want a previously used collection to hold a reference to objects that were assigned in the previous request.